src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPool.cs

```diff
@@ -2,130 +2,440 @@
+                _usedSinceLastCleanup = false;
             }
 
-            public bool TryAdd(HttpConnection connection)
+            // Pool is active.  Should not be removed.
+            return false;
+        }
+
+        /// <summary>A cached idle connection and metadata about it.</summary>
+        [StructLayout(LayoutKind.Auto)]
+        private struct CachedConnection : IEquatable<CachedConnection>
+        {
+            /// <summary>The cached connection.</summary>
+            internal readonly HttpConnection _connection;
+            /// <summary>The last time the connection was used.</summary>
+            internal readonly DateTimeOffset _returnedTime;
+
+            /// <summary>Initializes the cached connection and its associated metadata.</summary>
+            /// <param name="connection">The connection.</param>
+            public CachedConnection(HttpConnection connection)
             {
-                HashSet<HttpConnection> set = _sets[connection.GetHashCode() & s_slotsMask];
-                lock (set) return set.Add(connection);
+                _connection = connection;
+                _returnedTime = DateTimeOffset.UtcNow;
             }
 
-            public bool TryRemove(HttpConnection connection)
+            /// <summary>Gets whether the connection is currently usable.</summary>
+            /// <param name="now">The current time.  Passed in to ammortize the cost of calling DateTime.UtcNow.</param>
+            /// <returns>
+            /// true if we believe the connection can be reused; otherwise, false.  There is an inherent race condition here,
+            /// in that the server could terminate the connection or otherwise make it unusable immediately after we check it,
+            /// but there's not much difference between that and starting to use the connection and then having the server
+            /// terminate it, which would be considered a failure, so this race condition is largely benign and inherent to
+            /// the nature of connection pooling.
+            /// </returns>
+            public bool IsUsable(DateTimeOffset now) =>
+                now - _returnedTime <= TimeSpan.FromMilliseconds(MaxIdleTimeMilliseconds) &&
+                !_connection.ReadAheadCompleted;
+
+            public bool Equals(CachedConnection other) => ReferenceEquals(other._connection, _connection);
+            public override bool Equals(object obj) => obj is CachedConnection && Equals((CachedConnection)obj);
+            public override int GetHashCode() => _connection?.GetHashCode() ?? 0;
+        }
+
+        /// <summary>Provides a waiter object that supports a generic function and state type for creating connections.</summary>
+        private sealed class ConnectionWaiter<TState> : ConnectionWaiter
+        {
+            /// <summary>The function to invoke if/when <see cref="CreateConnectionAsync"/> is invoked.</summary>
+            private readonly Func<TState, ValueTask<HttpConnection>> _func;
```

Comment:
nit, can we call this something like _createConnection?  I think that's what it's called in the above code.